composer/core/state.py

```diff
@@ -89,15 +98,14 @@ def __init__(
 
             # data configurations
             grad_accum: int,
-            train_dataloader: types.DataLoader,
-            eval_dataloader: types.DataLoader,
+            train_data: Union[types.DataLoader, types.DataSpec],
```

Comment:
I would prefer to keep it as `train_dataloader` even though one can pass in a dataSpec

Reply:
From user standpoint, I think the desired interface would be to pass in either a `dataloader` or a config, which would be a `dict` with keys for the `device_transforms`, `batch_split`, etc.

Reply:
I will `DataSpec` to allow passing of function pointers in init. The API would thus look like:

`state = State(train_dataloader=DataSpec(num_tokens=100, device_transforms=lambda x: x, ...)`

We don't use dicts for configs anywhere since they aren't typed, but I think the above API is quite similar to that.

Reply:
Let's use dicts instead as that is the most light-weight datatype that accomplishes what we need. With the above approach, the user still needs to directly import `DataSpec` in their code to even create our `State`. So then `train_dataloader` would be either the pytorch dataloader or a `dict` config. From usability perspective, we want to replace these `Spec`-holding classes to native dictionary types as much as possible.

Reply:
Discussed offline that we'll leave this for now. I did update the constructor to allow for dicts so one doesn't have to import DataSpec directly.